drivers/adc/adc_nrfx_saadc.c

```diff
@@ -26,170 +24,242 @@ static struct driver_data m_data = {
-		config.acq_time = NRF_SAADC_ACQTIME_40US;
+		ch_cfg->acq_time = NRF_SAADC_ACQTIME_40US;
 		break;
 	default:
 		LOG_ERR("Selected ADC acquisition time is not valid");
 		return -EINVAL;
 	}
 
-	config.mode = (channel_cfg->differential ?
-		NRF_SAADC_MODE_DIFFERENTIAL : NRF_SAADC_MODE_SINGLE_ENDED);
+	return 0;
+}
 
-	/* Keep the channel disabled in hardware (set positive input to
-	 * NRF_SAADC_INPUT_DISABLED) until it is selected to be included
-	 * in a sampling sequence.
-	 */
+/* Implementation of the ADC driver API function: adc_channel_setup. */
+static int adc_nrfx_channel_setup(const struct device *dev,
+				  const struct adc_channel_cfg *channel_cfg)
+{
+	nrf_saadc_channel_config_t *ch_cfg;
+	nrfx_saadc_channel_t cfg = {
+		.channel_config = {
+			.resistor_p = NRF_SAADC_RESISTOR_DISABLED,
+			.resistor_n = NRF_SAADC_RESISTOR_DISABLED,
+			.burst      = NRF_SAADC_BURST_DISABLED,
+		},
+		.channel_index = channel_cfg->channel_id,
+	};
+	int err;
 
-	nrf_saadc_channel_init(NRF_SAADC, channel_id, &config);
-	nrf_saadc_channel_input_set(NRF_SAADC,
-				    channel_id,
-				    NRF_SAADC_INPUT_DISABLED,
-				    channel_cfg->input_negative);
+	if (channel_cfg->channel_id >= SAADC_CH_NUM) {
+		return -EINVAL;
+	}
 
-	/* Store the positive input selection in a dedicated array,
-	 * to get it later when the channel is selected for a sampling
-	 * and to mark the channel as configured (ready to be selected).
-	 */
-	m_data.positive_inputs[channel_id] = channel_cfg->input_positive;
+	ch_cfg = &cfg.channel_config;
+
+	err = pin_assign(channel_cfg, &cfg);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	err = gain_set(ch_cfg, channel_cfg->gain);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	err = reference_set(ch_cfg, channel_cfg->reference);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	err = acq_time_set(ch_cfg, channel_cfg->acquisition_time);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	ch_cfg->mode = (channel_cfg->differential ?
+		NRF_SAADC_MODE_DIFFERENTIAL : NRF_SAADC_MODE_SINGLE_ENDED);
+
+	nrfx_saadc_channel_config(&cfg);
 
 	return 0;
 }
 
 static void adc_context_start_sampling(struct adc_context *ctx)
 {
-	nrf_saadc_enable(NRF_SAADC);
+	nrfx_err_t err;
 
 	if (ctx->sequence.calibrate) {
-		nrf_saadc_task_trigger(NRF_SAADC,
-				       NRF_SAADC_TASK_CALIBRATEOFFSET);
-	} else {
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_START);
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_SAMPLE);
+		nrfx_saadc_offset_calibrate(NULL);
+	}
+
+	err = nrfx_saadc_mode_trigger();
+
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Cannot start sampling: %d", err);
 	}
 }
 
 static void adc_context_update_buffer_pointer(struct adc_context *ctx,
 					      bool repeat)
 {
+	nrf_saadc_value_t *buffer;
+	uint32_t buffer_size;
+
 	ARG_UNUSED(ctx);
 
-	if (!repeat) {
-		nrf_saadc_buffer_pointer_set(
-			NRF_SAADC,
-			nrf_saadc_buffer_pointer_get(NRF_SAADC) +
-			nrf_saadc_amount_get(NRF_SAADC));
+	if (repeat) {
```

Comment:
what about:
```
if (!repeat) {
		nrfx_saadc_buffer_set(m_data.event->data.done.p_buffer +
			 m_data.event->data.done.size, m_data.event->data.done.size);
}
```
for `repeat == true` no action is needed because driver preserves the previously configured buffer (for simple mode at least, but I see that advanced mode is not utilized by the shim)